File: LCA-go/lca-test.go
package lca
import "testing"

// Test basic LCA queries with binary tree.
func TestBasicTree(t *testing.T) {
	root := Node{1, nil, nil}
	root.left = &Node{2, nil, nil}
	root.right = &Node{3, nil, nil}
	root.left.left = &Node{4, nil, nil}
	root.left.right = &Node{5, nil, nil}
	root.right.left = &Node{6, nil, nil}
	root.right.right = &Node{7, nil, nil}

	tables := []struct {
        n1 int
        n2 int
        expectedLCA int
    }
	{
        {4, 5, 2},
        {4, 3, 1},
        {2, 3, 1},
        {2, 7, 1},
    }
	for _, table := range tables {
		lca := leastCommonAncestorInBST(&root, table.n1, table.n2).key
		if lca != table.expectedLCA {
			t.Errorf("LCA of %d and %d is incorrect, got %d, want %d.", table.n1, table.n2, lca, table.expectedLCA)
		}
	}
}

// Test straight binary tree.
func TestStraightTree(t *testing.T) {
	root := Node{1, nil, nil}
	root.right = &Node{2, nil, nil}
	root.right.right = &Node{3, nil, nil}
	root.right.right.right = &Node{4, nil, nil}
	lca := leastCommonAncestorInBST(&root, 4, 3).key
	if lca != 3 {
		t.Errorf("LCA is incorrect, got %d, want %d", lca, 3)
	}
}

// Test when one of the provided nodes is the LCA.
func TestNodeIsLCA(t *testing.T) {
	root := Node{1, nil, nil}
	root.left = &Node{2, nil, nil}
	root.right = &Node{3, nil, nil}
	root.left.left = &Node{4, nil, nil}
	root.left.right = &Node{5, nil, nil}
	lca := leastCommonAncestorInBST(&root, 4, 2).key
	if lca != 2 {
		t.Errorf("LCA is incorrect, got %d, want %d", lca, 2)
	}
}

// Test whe provided node is not in the binary tree.
func TestNodeNotIncluded(t *testing.T) {
	root := Node{1, nil, nil}
	root.left = &Node{2, nil, nil}
	root.right = &Node{3, nil, nil}
	root.left.left = &Node{4, nil, nil}
	root.left.right = &Node{5, nil, nil}
	lca := leastCommonAncestorInBST(&root, 0, 2)
	if lca != nil {
		t.Errorf("LCA is incorrect, should return nil if node(s) not present in tree")
	}
}

// Test when there is no binary tree (root is nil).
func TestNullTree(t *testing.T) {
	var root *Node = nil
	lca := leastCommonAncestorInBST(root, 1, 2)
	if lca != nil {
		t.Errorf("LCA is incorrect, empty tree should return nil")
	}
}

File: LCA-python/lca-test.py
import unittest
import lca

class LCATest(unittest.TestCase):

    # Test basic LCA queries with binary tree.
    def test_basic_tree(self):
        root = lca.Node(1)
        root.left = lca.Node(2)
        root.right = lca.Node(3)
        root.left.left = lca.Node(4)
        root.left.right = lca.Node(5)
        root.right.left = lca.Node(6)
        root.right.right = lca.Node(7)

        result1 = lca.findLCA(root, 4, 5).key
        result2 = lca.findLCA(root, 5, 3).key

        self.assertEqual(result1, 2)
        self.assertEqual(result2, 1)

    # Test straight binary tree.
    def test_straight_tree(self):
        root = lca.Node(1)
        root.right = lca.Node(2)
        root.right.right = lca.Node(3)
        root.right.right.right = lca.Node(4)

        result = lca.findLCA(root, 4, 3).key
        self.assertEqual(result, 3)

    # Test when one of the provided nodes is the LCA.
    def test_node_is_lca(self):
        root = lca.Node(1)
        root.left = lca.Node(2)
        root.right = lca.Node(3)
        root.left.left = lca.Node(4)
        root.left.right = lca.Node(5)

        result = lca.findLCA(root, 4, 2).key
        self.assertEqual(result, 2)

    # Test when provided node is not in the binary tree.
    def test_node_not_included(self):
        root = lca.Node(1)
        root.left = lca.Node(2)
        root.right = lca.Node(3)
        root.left.left = lca.Node(4)
        root.left.right = lca.Node(5)

        result = lca.findLCA(root, 0, 2)
        self.assertEqual(result, None)

    # Test when there is no binary tree (root is None).
    def test_null_tree(self):
        root = None
        result = lca.findLCA(root, 1, 2)
        self.assertEqual(result, None)

if __name__ == '__main__':
    unittest.main()